### PR TITLE
fix(tui): add classic input frame separator

### PR DESCRIPTION
**Pull request by jinyee8753**
## Summary
- Adds env-gated classic TUI status bar/input framing via `HERMES_TUI_CLASSIC_SKIN`.
- Preserves classic statusbar customization across CLI/TUI gateway paths.
- Frames the live composer prompt line with visible separators above and below in classic mode.
- Keeps default TUI behavior unchanged unless the classic-skin env flag is enabled.

## Visual smoke evidence

Captured from tmux after running:

```bash
HERMES_TUI_CLASSIC_SKIN=1 ./venv/bin/python hermes_cli/main.py --tui
```

Typed input smoke:

```text
56  ─ ready │ gpt 5.5 · medium │ 0/272k │ [░░░░░░░░░░] 0% │ ….hermes/hermes-agent (…lassic-followup) │ 3s │ voice off
57  ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
58  ❯ abc input smoke
59    ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

The separator is intentionally attached to the live composer prompt row, not just the status bar row.

## Test Plan
- [x] `cd ui-tui && npm run type-check`
- [x] `npx eslint src/config/env.ts src/components/appLayout.tsx src/components/appChrome.tsx src/app/slash/commands/session.ts --quiet`
- [x] `cd ui-tui && npm run build`
- [x] `./venv/bin/python -m pytest tests/cli/test_cli_status_bar.py -q -o 'addopts='` (`39 passed`)
- [x] tmux smoke with `HERMES_TUI_CLASSIC_SKIN=1` confirmed composer prompt separators above/below

## Notes
- Rebasing was done on current `origin/main` before opening this PR.
- This is ready for maintainer review; the visual change remains opt-in behind the classic-skin env flag.

### CODE REVIEW

**Comment by cardtest15-coder** [comment]
The user wants a code review of a PR in Russian. The PR touches multiple files: cli.py, hermes_cli/main.py, tests, tui_gateway/server.py, and ui-tui components.

Let me analyze the diff carefully.

**1. cli.py changes:**
- `_reasoning_effort_label`: Returns "medium" if None, "none" if dict with enabled False, otherwise effort or "medium". If not dict, returns "". Wait, if reasoning_config is a string? It returns "". That seems okay but maybe could be stricter.
- `_status_bar_model_label`: Gets reasoning_config from agent or self. Gets service_tier from agent or self. Builds parts list. If effort_label is truthy, adds it. If service_tier == "priority", adds "fast". Joins with " · ".
  - Bug potential: `effort_label` could be empty string `""` if reasoning_config is not a dict. But the condition `if effort_label:` handles that. However, if reasoning_config is `{"enabled": True, "effort": ""}`, it returns "medium" because `or "medium"`. That's fine.
  - Wait: `_reasoning_effort_label` returns "medium" for None. But `_status_bar_model_label` only adds effort_label if it's truthy. "medium" is truthy, so it will always show "medium" if reasoning_config is None? That seems intentional but maybe noisy. The PR seems to want to show it.
- `_status_bar_workspace_label`: Uses `TERMINAL_CWD` env var or `os.getcwd()`. Tries Path(cwd).expanduser(). If path == home, returns "~". Else returns path.name or str(path). Exception fallback splits by "/" and takes last part.
  - Issue: The fallback `str(cwd).rstrip("/\\").split("/")[-1]` uses `/` as separator but on Windows backslash paths won't split correctly by `/`. However `str(cwd).rstrip("/\\").split("/")[-1]` - if Windows path is `C:\Users\foo`, splitting by "/" gives `["C:\Users\foo"]`, so `[-1]` is the whole thing. Not great but acceptable as fallback. Could use `os.path.basename` or `os.path.split` but it's a fallback.
  - Issue: `path.name or str(path)` - if path is root `/`, `path.name` is empty string? Actually `Path('/').name` returns `''` (empty). So it returns `str(path)` which is `/`. That's okay. If path is `~`, expanduser makes it `/home/user`, name is `user`. Good.
- `_get_status_bar_snapshot`: Adds chat_status, model_label, workspace_label. `chat_status` checks `getattr(self, "_agent_running", False)`. This seems okay.
- `_build_status_bar_text` and `_get_status_bar_fragments`: Modified to include chat_status, model_label, workspace_label.
  - In `_build_status_bar_text` for width < 76: `compact_model = self._status_bar_model_label(self._trim_status_bar_text(snapshot["model_short"], 12))`. Wait, `_trim_status_bar_text` is designed to trim a full status bar text string, not just a model short label. It might trim aggressively or add "..." in weird ways. But it's used here to trim the model short before passing to `_status_bar_model_label`. Actually `_trim_status_bar_text` takes `text` and `max_width`. It calculates visible width. If it trims, it adds "...". But passing a 12-char max width to a model short like "claude-sonnet-4-20250514" (which is already truncated to 26 chars in snapshot) and then trimming to 12... seems okay but `_trim_status_bar_text` might not be the right tool for simple string truncation. However, it's static and works. But wait: `_trim_status_bar_text` uses `self._visible_width`? No, it's not shown as static in the diff, but likely is an instance method. Actually in the diff it's called as `self._trim_status_bar_text`. It might compute visible length of emoji etc. It's okay.
  - In `_get_status_bar_fragments` for width < 76: same compact_model logic.
  - In `_get_status_bar_fragments` wide mode: uses `snapshot["model_label"]` directly.
  - Issue: In `_build_status_bar_text` for width < 52: `text = f"{snapshot['chat_status']} │ ⚕ {snapshot['model_label']} · {duration_label}"`. But `snapshot['model_label']` already contains model_short + effort + fast. If model_short was already truncated to 26 in snapshot, and then model_label adds effort and fast, it could exceed width. But `_trim_status_bar_text` is called at return. Okay.
  - Issue: `snapshot["workspace_label"]` is included in wide mode but not in narrow mode (<76). That's intentional.

**2. hermes_cli/main.py changes:**
- `_run_local_post_update_hooks`: Runs shell scripts from `HERMES_HOME/local-patches/hermes-statusbar/apply.sh` or fallback `~/.hermes/...`.
  - **Security issue**: It runs shell scripts via `subprocess.run([str(hook)])` without shell=True, so it's executing the script directly. But the script could be anything. However, it's user-owned hooks under HERMES_HOME. The comment says "user-owned post-update hooks that live outside the source tree". This is a deliberate feature, but there are some concerns:
    - No validation of hook file permissions (is it writable by others?).
    - `cwd=PROJECT_ROOT` is used.
    - It's best-effort and exceptions are caught.
    - Not a major security vulnerability if it's intentional and documented, but worth noting.
  - Wait: `subprocess.run([str(hook)], cwd=PROJECT_ROOT, text=True)` - no `check=True`, so return code is just printed. Good.
  - The function is called in `cmd_update` after configuration update. Seems fine.

**3. tests/cli/test_cli_status_bar.py:**
- Added tests for wide terminal, reasoning effort, reasoning none, working status.
- They patch `os.environ` for TERMINAL_CWD.
- They check string contents.
- Missing tests: No tests for `_status_bar_workspace_label` fallback when `TERMINAL_CWD` is not set. No tests for `_reasoning_effort_label` edge cases (e.g., reasoning_config is a string, or dict without "enabled" key). No tests for `_get_status_bar_fragments` narrow width (<52) or medium width (<76) with the new fields. The existing tests seem to cover wide terminal and some specific features.
- The test `test_build_status_bar_text_for_wide_terminal` now patches TERMINAL_CWD. Good.
- The test `test_compression_count_in_wide_fragments` patches `_get_tui_terminal_width` to 120. Good.

**4. tui_gateway/server.py:**
- Changes to `/set` command handlers for `fast` and `reasoning`.
- For `fast`: now returns `result` which includes `{"key": key, "value": nv}` and if agent exists, also `info` from `_session_info(agent)`. Also emits `session.info` event.
- For `reasoning`: similar pattern. When `show` or `hide`, returns result with info if agent exists. When setting effort (parsed), updates agent.reasoning_config, emits session.info, and returns result with info.
- Logic correctness: The `if arg in {"hide", "off"}` was changed to `if arg in ("hide", "off")`. Fine.
- The `result` dict is built and returned. This helps UI stay in sync. Good.
- Potential bug: In the `fast` handler, `_write_config_key("agent.service_tier", nv)` is called before checking if `nv` is valid? Actually `nv` is derived from `value` (fast/normal). It's validated above.
- In reasoning handler, `_write_config_key("agent.reasoning_effort", arg)` is called before checking if agent exists. That's fine, it's config. Then if agent exists, updates `agent.reasoning_config = parsed`. Good.

**5. ui-tui changes:**
- `session.ts`: For reasoning command, if `r.info`, patches UI state info. For fast command, if `r.info`, patches UI state info, else falls back to manual patch. Good.
- `appChrome.tsx`:
  - `effortLabel`: Changed condition from `value !== 'medium' && value !== 'normal' && value !== 'default'` to `value !== 'normal' && value !== 'default'`. So now "medium" is shown? Wait, previously medium was hidden. Now medium is shown? Let's check: `return value && value !== 'normal' && value !== 'default' ? value : ''`. So if value is "medium", it returns "medium". That means the status bar will show "medium". But in the backend `_reasoning_effort_label` returns "medium" as default. So the UI now shows it. This aligns with backend changes.
  - `fastLabel`: returns 'fast' if true else ''.
  - `modelLabel`: Joins with ` · ` instead of ` `. Aligns with backend.
  - `GoodVibesHeart`: Dependency array changed from `[t.color.accent, tick]` to `[t.color.accent, t.color.error, t.color.warn, tick]`. Why? The function uses `t.color.accent` only. Wait, looking at the code:
    ```tsx
    useEffect(() => {
      setActive(true)
      const id = setTimeout(() => setActive(false), 650)
      return () => clearTimeout(id)
    }, [t.color.accent, t.color.error, t.color.warn, tick])
    ```
    The effect only depends on `tick` to re-trigger. Adding theme colors to deps means if theme changes while active, it resets. But the colors aren't used in the effect body except maybe implicitly? Actually the component returns a heart emoji with color based on some condition not shown in diff? Wait, the component is truncated in diff? No, the diff shows:
    ```tsx
    export function GoodVibesHeart({ tick, t }: { tick: number; t: Theme }) {
      const [active, setActive] = useState(false)
      useEffect(() => {
        setActive(true)
        const id = setTimeout(() => setActive(false), 650)
        return () => clearTimeout(id)
      }, [t.color.accent, t.color.error, t.color.warn, tick])
    ```
    Wait, the effect doesn't use `t.color.error` or `t.color.warn`. It just sets active. This looks like a mistake or leftover from a previous iteration where the color was used. Or maybe the linter forced them because they are used in the JSX below? But the effect doesn't use them. Adding unused dependencies to useEffect will cause unnecessary re-runs. However, `tick` changes frequently anyway, so the effect runs often. The extra deps are harmless but slightly odd. Not a bug, just unnecessary.
  - `StatusRule`: Major restructuring.
    - `leftWidth = Math.max(12, cols)` instead of `cols - cwdLabel.length - 3`. Now cwdLabel is appended at the end of the left box, not in a separate right box. The separator logic changed.
    - `chatStatus` added.
    - Layout: `'─ '`, then chatStatus, then busy face ticker, then model label, ctxLabel, bar, cwdLabel, session time, cost, etc.
    - The `cwdLabel` is now inside the left `<Text>` instead of a separate `<Text>` after `<Box>`. This is better for truncation.
    - Wait, the old code had:
      ```tsx
      <Box ...>
        <Text ...> ... </Text>
      </Box>
      <Text color={t.color.border}> ─ </Text>
      <Text color={t.color.label}>{cwdLabel}</Text>
      ```
      New code puts cwdLabel inside the first Text. But the closing `</Box>` is still there. Wait, looking at the diff:
      ```diff
      -      <Text color={t.color.border}> ─ </Text>
      -      <Text color={t.color.label}>{cwdLabel}</Text>
         </Box>
      ```
      So the two trailing Text elements are removed, and cwdLabel is moved inside. Good.
    - But `leftWidth` is now `Math.max(12, cols)`. Since cwdLabel is inside the same Box with wrap="truncate-end", it will truncate the whole thing including cwdLabel if too long. That's correct.
    - However, there is a subtle issue: the `FaceTicker` is now wrapped in `<Text color={statusColor}> <FaceTicker ... /></Text>`. The space before FaceTicker is inside the Text. That's fine.
    - The `chatStatus` is rendered as `<Text color={statusColor}>{chatStatus}</Text>` but it's nested inside the outer `<Text color={t.color.border}>`. In Ink, nested Text components can override color. This should work.
    - Wait, the structure is:
      ```tsx
      <Text color={t.color.border} wrap="truncate-end">
        {'─ '}
        <Text color={statusColor}>{chatStatus}</Text>
        {busy ? (
          <Text color={statusColor}> <FaceTicker ... /></Text>
        ) : null}
        <Text color={t.color.muted}> │ {modelLabel(...)}</Text>
        ...
      </Text>
      ```
      This is valid Ink JSX.
- `appLayout.tsx`:
  - `CLASSIC_SKIN` env var added.
  - `horizontalRule` function repeats `─` for cols-2.
  - `TranscriptPane`: User message separator changed from `───` to `horizontalRule(composer.cols)`. That's a full-width rule. Good.
  - `ComposerPane`:
    - If `CLASSIC_SKIN && ui.statusBar === 'top'`, the spacer box is null (height 1 box removed). This avoids double spacing.
    - Added `horizontalRule` above input when CLASSIC_SKIN and statusBar top.
    - Input box height increased by 1 when CLASSIC_SKIN and composer.empty.
    - Placeholder changed for CLASSIC_SKIN to `\n${horizontalRule(inputColumns)}`. This is a multiline placeholder showing a horizontal rule.
    - When CLASSIC_SKIN and input.length > 0, a horizontal rule is shown below the input.
    - Wait: The placeholder is `\n${horizontalRule(inputColumns)}`. If the input is empty, the TextInput shows this placeholder. But TextInput might not handle newlines well? In Ink, TextInput with multiline placeholder? The component might support it. But it's a bit hacky.
    - The `horizontalRule(inputColumns)` uses `inputColumns` which is the width of the input box. But the rule is drawn inside the input box's container, not inside TextInput itself. Wait, looking at the code:
      ```tsx
      <Box flexDirection="column" ... height={CLASSIC_SKIN ? inputHeight + 1 : inputHeight} width={inputColumns}>
        <TextInput
          columns={inputColumns}
          ...
          placeholder={CLASSIC_SKIN ? `\n${horizontalRule(inputColumns)}` : ...}
          ...
        />
        {CLASSIC_SKIN && composer.input.length > 0 && (
          <Text color={ui.theme.color.border} wrap="truncate-end">
            {horizontalRule(inputColumns)}
          </Text>
        )}
      </Box>
      ```
      So when CLASSIC_SKIN is true and input is empty, the placeholder is a newline + rule. The TextInput component likely renders the placeholder string. If it handles `\n`, it will show the rule on the second line. But the Box height is `inputHeight + 1` when CLASSIC_SKIN and composer.empty. So there's room.
      When input.length > 0, a Text component below TextInput renders the rule. This effectively frames the input area with horizontal rules (one above from the previous Text, one below from this Text). This matches the PR title "add classic input frame separator".
    - However, the placeholder `\n${horizontalRule(inputColumns)}` might be problematic if TextInput doesn't expect newlines. It might wrap or truncate. But since it's their custom component, maybe it's fine.
    - `StatusRulePane`: `marginTop={at === 'top' ? 1 : 0}` changed to `marginTop={0}`. This removes top margin when status bar is at top. Combined with CLASSIC_SKIN, this tightens layout. But if CLASSIC_SKIN is false and statusBar is top, marginTop becomes 0 too. Is that intentional? The diff shows unconditional change. It might affect non-classic skin layout slightly. But maybe it's a fix.

**Potential bugs / issues:**

1. **cli.py**: `_status_bar_workspace_label` fallback uses `str(cwd).rstrip("/\\").split("/")[-1]`. On Windows, `os.getcwd()` might return `C:\Users\name`. `rstrip("/\\")` removes trailing backslashes. `split("/")` won't split on backslash, so `[-1]` returns the whole string `C:\Users\name`. That's not a great fallback. Should use `os.path.basename` or `pathlib` consistently. Minor.

2. **cli.py**: `_build_status_bar_text` for width < 76 uses `self._status_bar_model_label(self._trim_status_bar_text(snapshot["model_short"], 12))`. `_trim_status_bar_text` is likely meant for full status bar strings and might account for emoji widths etc. Using it to trim a model short string is okay but could be overkill. However, if `_trim_status_bar_text` uses ANSI stripping or width calculation, it should be fine. But what if `snapshot["model_short"]` is already <=12? It returns it as-is. Good.

3. **cli.py**: `_get_status_bar_fragments` for width < 76 uses `compact_model` similarly. But in the fragments, it uses `compact_model` which is computed from `snapshot["model_short"]` not `snapshot["model_label"]`. This means in medium width, the model label might not include reasoning/fast if the model short is trimmed? Wait, `_status_bar_model_label` takes a `model_short` string and appends reasoning/fast. So even if model_short is trimmed, the label is built from the trimmed short + effort + fast. That's correct.

4. **ui-tui/appChrome.tsx**: `leftWidth = Math.max(12, cols)`. Previously it was `Math.max(12, cols - cwdLabel.length - 3)`. Now cwdLabel is inside the truncating text. If cwdLabel is long, it will just be truncated along with everything else. But since it's at the end, truncation will cut off cwdLabel first (or earlier parts depending on space). Actually `wrap="truncate-end"` truncates at the end. So if the whole string is too long, the end (cwdLabel, session time, etc.) gets truncated. That's acceptable.

5. **ui-tui/appLayout.tsx**: `horizontalRule(com

**Comment by cardtest15-coder** [comment]
ERROR: LLM call failed: 412 Client Error: Precondition Failed for url: https://api.fireworks.ai/inference/v1/chat/completions

**Comment by teknium1** [comment]
Thanks for the contribution — I verified the underlying UI gap still exists on current main, but I would not salvage this PR as-is.

### Problems
- `ui-tui/src/config/env.ts:14` on PR head `16b07ff` adds `HERMES_TUI_CLASSIC_SKIN` as a user-facing display toggle. `AGENTS.md:102-106` says non-secret behavioral/display settings belong in `config.yaml`, not new `HERMES_*` env vars.
- `hermes_cli/main.py:7693` adds `_run_local_post_update_hooks()` and `hermes_cli/main.py:8249` runs it during update. That executable local-patch hook is unrelated to the TUI input-frame change and should not ride in this cosmetic TUI PR.
- The PR says default behavior is unchanged unless classic skin is enabled, but PR head `ui-tui/src/components/appLayout.tsx:381` changes `StatusRulePane` spacing unconditionally. Current main still uses `marginTop={at === 'top' ? 1 : 0}` at `ui-tui/src/components/appLayout.tsx:362`.

### Suggested changes
- Move the classic-frame toggle into normal display config / gateway config sync instead of an env var.
- Remove the update-hook changes from this PR.
- Gate spacing changes so the default TUI layout remains byte-for-byte behaviorally unchanged when the frame is disabled.

This is an automated hermes-sweeper review.

File: ui-tui/src/config/env.ts [hunk 1]
```diff
@@ -11,6 +11,7 @@ export const NO_CONFIRM_DESTRUCTIVE = truthy(process.env.HERMES_TUI_NO_CONFIRM)
 // Experiment gate: lets us measure native scroll vs our virtualization on
 // the same pipeline.
 export const INLINE_MODE = truthy(process.env.HERMES_TUI_INLINE)
+export const CLASSIC_SKIN = truthy(process.env.HERMES_TUI_CLASSIC_SKIN)
```
Comment (teknium1) [inline ui-tui/src/config/env.ts:14]: This is a user-facing display preference. Please route it through `display.*` config and the existing TUI config-sync path rather than introducing a new `HERMES_*` environment variable; `AGENTS.md:102-106` reserves `.env` for secrets.

File: hermes_cli/main.py [hunk 2]
```diff
@@ -7690,6 +7690,41 @@ def _run_pre_update_backup(args) -> None:
     print()
 
 
+def _run_local_post_update_hooks() -> None:
```
Comment (teknium1) [inline hermes_cli/main.py:7693]: Please remove this update-time executable hook from the TUI framing PR. It changes the `hermes update` execution path but is unrelated to rendering the composer frame.

File: ui-tui/src/components/appLayout.tsx [hunk 3]
```diff
@@ -349,7 +378,7 @@ const StatusRulePane = memo(function StatusRulePane({
   }
 
   return (
-    <Box marginTop={at === 'top' ? 1 : 0}>
+    <Box marginTop={0}>
```
Comment (teknium1) [inline ui-tui/src/components/appLayout.tsx:381]: This removes the top status-rule spacing even when `CLASSIC_SKIN` is false. Gate this layout change with the frame preference so disabled mode preserves current behavior.

**Review by teknium1 — COMMENTED** [review]
Thanks for the visual framing work. The composer-frame gap still exists on current main (`ui-tui/src/components/appLayout.tsx:404-416`), but this implementation needs rework before it can be salvaged.

### Problems
- `ui-tui/src/config/env.ts:14` adds `HERMES_TUI_CLASSIC_SKIN` as a user-facing display toggle. `AGENTS.md:102-106` requires display preferences to use `config.yaml`, not a new `HERMES_*` environment variable.
- `hermes_cli/main.py:7693` adds an executable post-update local hook and `hermes_cli/main.py:8249` runs it during updates. That behavior is unrelated to composer framing and should not be included here.
- `ui-tui/src/components/appLayout.tsx:381` removes the top status-rule margin unconditionally. Current main still uses `marginTop={at === 'top' ? 1 : 0}` at `ui-tui/src/components/appLayout.tsx:467`, so the default layout changes when the frame is disabled.
- The diff adds no TUI regression test for the frame flag or its composer/status-bar state combinations.

### Suggested changes
- Use a `display.*` config setting and the existing TUI config-sync path (`ui-tui/src/app/useConfigSync.ts:219-233`).
- Remove the update-hook change, gate all layout changes, and add focused TUI coverage for enabled and disabled states.

This is an automated hermes-sweeper review.
